### PR TITLE
Semipersistent Security Records

### PR DESCRIPTION
**Pull request by ike709**
## About The Pull Request

I didn't feel like doing any of the UI work for crosscargo so I did this PR instead. A player's criminal record will now persist for a configurable number of days. It's tied to character name unless the player randomizes their name, then it's tied to ckey.

Yes, antags are ignored. Also I have a terrible hangover so don't roast me too hard if I wrote bad code.

Also crime datums don't seem to ever be deleted. Now they are.

## Why It's Good For The Game

Currently, security has pretty minimal incentive to actually use security records properly, because the round ends after an hour or two. This enables security to actually keep track of criminal behavior across rounds, without having an endlessly growing persistent track record. The actual implications of this feature will be decided by policy which hasn't been discussed yet, so the feature will be disabled by default.

## TODO

- [x] Get Crossed to help with making bans auto-expire on the DB side
- [x] Finalize the SQL after going over it with Crossed
- [x] Write the SQL migration changelog thing.
- [x] Check if MassInsert sanitizes and manually sanitize if it doesn't.
- [x] Test it
- [ ] Disable it in config until policy can be written.

## Changelog
:cl:
add: Security records will now persist between rounds for 7 days (CURRENTLY DISABLED IN CONFIG).
This does not apply to antags. It will attempt to tie your record to your character unless you use a random name, in which case it is tied to ckey. Deleting your record does in fact remove it from the DB.
/:cl:

### CODE REVIEW

**Comment by ghost** [comment]
>then it's tied to ckey

That's going to be weird to RP having aggregated crimes for two different characters.

**Comment by francinum** [comment]
> > then it's tied to ckey
> 
> That's going to be weird to RP having aggregated crimes for two different characters.

It's the best option for dealing with always-random characters

**Comment by ghost** [comment]
This seems interesting in concept, I suspect that this is intended to deal with repeat offenders?

**Comment by MNarath1** [comment]
I think antags should be excluded from this like a crime record you got during a round as an antagonist should not persist as your behavior was most likely part of your mission as said particular antag.

**Comment by ike709** [comment]
> I think antags should be excluded from this like a crime record you got during a round as an antagonist should not persist as your behavior was most likely part of your mission as said particular antag.

Please get into the habit of reading a PR description before commenting on the PR. @MNarath1

**Comment by MNarath1** [comment]
> > I think antags should be excluded from this like a crime record you got during a round as an antagonist should not persist as your behavior was most likely part of your mission as said particular antag.
> 
> 
> 
> Please get into the habit of reading a PR description before commenting on the PR. @MNarath1 

Ah shit i totally overread that sorry.

**Comment by SergeiKoralev** [comment]
Can we also generate a crimes.log that populates with when people are set to wanted as well as when are crimes added? This would be amazingly helpful for finding shitcurity for admins.

**Comment by ike709** [comment]
Note to self: Have deleting a record remove it from the DB.

**Comment by ItsTisme** [comment]
The clown deffintly will have fun with this

**Comment by eeSPee** [comment]
Metagrudge moment?

> because the round ends after an hour or two

Xenobiology shrugs.

**Comment by ike709** [comment]
It looks like data datums were never deleted when removed from a record or the record itself was deleted. This should now be fixed, but I don't know data core code well enough to know if this has unintended side-effects.

File: SQL/beestation_schema.sql [hunk 1]
```diff
@@ -546,6 +546,21 @@ END
...
+  `ckey` varchar(32) NOT NULL,
+  `crime` varchar(2048) NOT NULL,
+  `details` varchar(2048) NOT NULL,
+  `author` varchar(32) NOT NULL,
+  `time` datetime NOT NULL,
+  `fine` int(10) unsigned NOT NULL,
+  `author_ckey` varchar(32) NOT NULL,
+  `paid` int(10) unsigned NOT NULL,
+  `character_name` varchar(32) NOT NULL,
+  PRIMARY KEY (`ckey`)
```
Comment (Crossedfall) [inline SQL/beestation_schema.sql]: Using the ckey as the primary key is going to cause you a lot of problems. At least if you want the ability to have multiple records tied to the same key.

File: code/datums/datacore.dm [hunk 2]
```diff
@@ -291,6 +328,46 @@
...
 	return
 
+/datum/datacore/proc/generate_security_record(id, mob/living/carbon/human/H)
+	var/datum/data/record/S = new()
+	S.fields["ckey"]		= H.ckey
+	S.fields["id"]			= id
+	S.fields["name"]		= H.real_name
+	S.fields["criminal"]	= "None"
+	S.fields["citation"]	= list()
+	S.fields["crim"]		= list()
```
Comment (Crossedfall) [inline code/datums/datacore.dm:338]: Is this a typo? "Crim"

File: code/datums/datacore.dm [hunk 3]
```diff
@@ -291,6 +328,46 @@
...
+	var/list/data_to_upload = list()
+	for(var/datum/data/record/S as() in GLOB.data_core.security)
+		if(!S.fields["ckey"] || !S.fields["name"])
+			continue
+		var/list/crimes = S.fields["crim"]
+		for(var/datum/data/crime/C in crimes)
+			if(C.fromDB || C.had_special_role || C.fine > 0) //We don't want citations, crimes committed while antag, and we don't want to reupload existing crimes
+				continue
+
+			data_to_upload += list("ckey" = S.fields["ckey"], "crime" = C.crimeName, "details" = C.crimeDetails, "author" = C.author, "author_ckey" = C.authorCkey, "character_name" = S.fields["name"])
```
Comment (Crossedfall) [inline code/datums/datacore.dm]: `MassInsert` requires nested lists for each row.
```suggestion
			data_to_upload += list(list("ckey" = S.fields["ckey"], "crime" = C.crimeName, "details" = C.crimeDetails, "author" = C.author, "author_ckey" = C.authorCkey, "character_name" = S.fields["name"]))
```

File: SQL/beestation_schema.sql [hunk 4]
```diff
@@ -546,6 +546,21 @@ END
...
+CREATE TABLE IF NOT EXISTS `SS13_criminal_records` (
+  `ckey` varchar(32) NOT NULL,
+  `crime` varchar(2048) NOT NULL,
+  `details` varchar(2048) NOT NULL,
+  `author` varchar(32) NOT NULL,
+  `time` datetime NOT NULL,
+  `fine` int(10) unsigned NOT NULL,
+  `author_ckey` varchar(32) NOT NULL,
+  `paid` int(10) unsigned NOT NULL,
+  `character_name` varchar(32) NOT NULL,
```
Comment (Crossedfall) [inline SQL/beestation_schema.sql]: Should add a couple of defaults in to make things a bit less error prone. Also change the `time` column to something a bit more descriptive that'd also prevent sql conflicts
```suggestion
  `ckey` varchar(32) NOT NULL,
  `crime` varchar(2048) DEFAULT NULL,
  `details` varchar(2048) DEFAULT NULL,
  `author` varchar(32) NOT NULL,
  `recordtime` datetime NOT NULL DEFAULT current_timestamp(),
  `fine` int(10) unsigned NOT NULL DEFAULT 0,
  `author_ckey` varchar(32) NOT NULL,
  `paid` int(10) unsigned NOT NULL DEFAULT 0,
  `character_name` varchar(32) NOT NULL,
```

**Review by Crossedfall — CHANGES_REQUESTED** [review]

File: code/datums/datacore.dm (hunk 2 above)
Comment (ike709) [reply code/datums/datacore.dm:338]: Surprisingly, it is not.

File: code/datums/datacore.dm [hunk 5]
```diff
@@ -291,6 +328,46 @@
 		locked += L
 	return
 
+/datum/datacore/proc/generate_security_record(id, mob/living/carbon/human/H)
+	var/datum/data/record/S = new()
+	S.fields["ckey"]		= H.ckey
```
Comment (ike709) [inline code/datums/datacore.dm]:
```suggestion
	S.fields["ckey"]		= H.client.ckey
```

File: code/datums/datacore.dm [hunk 6]
```diff
@@ -265,14 +308,8 @@
...
-		var/datum/data/record/S = new()
-		S.fields["id"]			= id
-		S.fields["name"]		= H.real_name
-		S.fields["criminal"]	= "None"
-		S.fields["citation"]	= list()
-		S.fields["crim"]		= list()
-		S.fields["notes"]		= "No notes."
-		security += S
+
+		security += generate_security_record(id, H)
```
Comment (ike709) [inline code/datums/datacore.dm]:
```suggestion
		security += generate_security_record(id, H, C.ckey)
```

File: code/datums/datacore.dm [hunk 7]
```diff
@@ -291,6 +328,46 @@
 		locked += L
 	return
 
+/datum/datacore/proc/generate_security_record(id, mob/living/carbon/human/H)
```
Comment (ike709) [inline code/datums/datacore.dm]:
```suggestion
/datum/datacore/proc/generate_security_record(id, mob/living/carbon/human/H, client/C)
```

File: code/datums/datacore.dm [hunk 8]
```diff
@@ -265,14 +308,8 @@
...
-		var/datum/data/record/S = new()
-		S.fields["id"]			= id
-		S.fields["name"]		= H.real_name
-		S.fields["criminal"]	= "None"
-		S.fields["citation"]	= list()
-		S.fields["crim"]		= list()
-		S.fields["notes"]		= "No notes."
-		security += S
+
+		security += generate_security_record(id, H, C.ckey)
```
Comment (ike709) [inline code/datums/datacore.dm]:
```suggestion
		security += generate_security_record(id, H, C)
```

File: code/datums/datacore.dm [hunk 9]
```diff
@@ -291,6 +328,46 @@
 		locked += L
 	return
 
+/datum/datacore/proc/generate_security_record(id, mob/living/carbon/human/H)
+	var/datum/data/record/S = new()
+	S.fields["ckey"]		= H.client.ckey
```
Comment (ike709) [inline code/datums/datacore.dm]:
```suggestion
	S.fields["ckey"]		= C.ckey
```

File: code/datums/datacore.dm [hunk 10]
```diff
@@ -291,6 +328,46 @@
...
+	S.fields["crim"]		= list()
+	S.fields["notes"]		= "No notes."
+	if(CONFIG_GET(flag/persist_security_records) && SSdbcore.Connect())
+		var/datum/DBQuery/crime_record = SSdbcore.NewQuery(
+			"SELECT crime, details, author, author_ckey, character_name FROM [format_table_name("criminal_records")] WHERE ckey = :ckey",
+			list("ckey" = H.ckey)
+		)
+		if(crime_record.Execute(async = TRUE))
+			while(crime_record.NextRow())
+				if(!H.client.prefs.be_random_name && (H.real_name != crime_record.item[5] && reject_bad_name(crime_record.item[5]) != null)) // Tie the records to the character unless they're using random names or the name isn't valid.
```
Comment (ike709) [inline code/datums/datacore.dm]:
```suggestion
				if(!C.prefs.be_random_name && (H.real_name != crime_record.item[5] && reject_bad_name(crime_record.item[5]) != null)) // Tie the records to the character unless they're using random names or the name isn't valid.
```

**Comment by ike709** [comment]
Survives basic testing.

TODO: Scrub the author from the record. Also make it more obvious that it is special by doing some text formatting.

File: code/datums/datacore.dm (hunk 2 above)
Comment (Archanial) [reply code/datums/datacore.dm:338]: Originally, crimes were split into major and minor, and those were called mi_crim and ma_crim. I just left it as crim.

**Comment by ike709** [comment]
I'm declaring this ready for review and testmerge. Crossed will need to tweak the DB/config for the testmerge to work.

**Comment by ghost** [comment]
This is cool, it really is. But, in a theoretical situation where it is enabled, what is its actual purpose? 

It is _technically_ against the rules to use previous round information (please don't send in the rule lawyer armada, I put technically there for a reason), and other than that it's just... there? How would a new policy as you mentioned circumnavigate such rules?

**Comment by WhyIsCaeciliusTaken** [comment]
> Can we also generate a crimes.log that populates with when people are set to wanted as well as when are crimes added? This would be amazingly helpful for finding shitcurity for admins.

This already exists, it is called record.log

**Comment by BettonnCZ** [comment]
I'm 90% sure this will end up being a reason to metagrudge. This MRP bruv, other round is a **new shift**

**Comment by BettonnCZ** [comment]
what purpose does this have, does it enhance the gameplay in a good way? People are gonna take advantage of that and grudge your ass to the point its not fun

**Pull request closed without merging**